velox/exec/tests/HashJoinTest.cpp

```diff
@@ -43,48 +69,184 @@ class HashJoinTest : public HiveConnectorTestBase {
     // memory usage for the hash join node is > 0.
     auto planStats = toPlanStats(task->taskStats());
     auto joinNodeId = params.planNode->id();
     ASSERT_TRUE(planStats.at(joinNodeId).peakMemoryBytes > 0);
   }
 
+  void testJoin(
+      const std::vector<TypePtr>& keyTypes,
+      int numThreads,
+      int leftBatchSize,
+      int numLeftBatches,
+      int rightBatchSize,
+      int numRightBatches,
+      const std::string& referenceQuery,
+      const std::string& filter = "",
+      const std::vector<std::string>& outputLayout = {},
+      core::JoinType joinType = core::JoinType::kInner) {
+    auto leftType = makeRowType(keyTypes, "t_");
+    auto rightType = makeRowType(keyTypes, "u_");
+
+    std::vector<RowVectorPtr> leftBatches;
+    std::vector<RowVectorPtr> rightBatches;
+    auto opts = fuzzerOpts_;
+    opts.vectorSize = leftBatchSize;
+    fuzzer_.setOptions(opts);
+    for (int i = 0; i < numLeftBatches; ++i) {
+      leftBatches.push_back(fuzzer_.fuzzRow(leftType));
+    }
+
+    opts.vectorSize = rightBatchSize;
+    fuzzer_.setOptions(opts);
+    for (int i = 0; i < numRightBatches; ++i) {
+      rightBatches.push_back(fuzzer_.fuzzRow(rightType));
+    }
+
+    testJoin(
+        numThreads,
+        leftType,
+        makeKeyNames(keyTypes.size(), "t_"),
+        leftBatches,
+        rightType,
+        makeKeyNames(keyTypes.size(), "u_"),
+        rightBatches,
+        referenceQuery,
+        filter,
+        outputLayout,
+        joinType);
+  }
+
+  void testJoin(
+      int32_t numThreads,
+      RowTypePtr leftType,
+      const std::vector<std::string>& leftKeys,
+      std::vector<RowVectorPtr>& leftBatches,
+      RowTypePtr rightType,
+      const std::vector<std::string>& rightKeys,
+      std::vector<RowVectorPtr>& rightBatches,
+      const std::string& referenceQuery,
+      const std::string& filter = "",
+      const std::vector<std::string>& outputLayout = {},
+      core::JoinType joinType = core::JoinType::kInner) {
+    // NOTE: there is one value node copy per driver thread and if the value
+    // node is not parallelizable, then the associated driver pipeline will be
+    // single threaded. We will try to test different multithreading
+    // combinations of build and probe sides, so populate duckdb with all the
+    // data: 'allLeftBatches' and 'allRightBatches'.
+    std::vector<RowVectorPtr> allLeftBatches;
+    allLeftBatches.reserve(leftBatches.size() * numThreads);
+    std::vector<RowVectorPtr> allRightBatches;
+    allLeftBatches.reserve(rightBatches.size() * numThreads);
+    for (int i = 0; i < numThreads; ++i) {
+      std::copy(
+          leftBatches.begin(),
+          leftBatches.end(),
+          std::back_inserter(allLeftBatches));
+      std::copy(
+          rightBatches.begin(),
+          rightBatches.end(),
+          std::back_inserter(allRightBatches));
+    }
+
+    createDuckDbTable("t", leftBatches);
+    createDuckDbTable("u", rightBatches);
+
+    {
```

Comment:
Can we put the three cases in a loop using pairs [{numDrivers, numDrivers}, {1, numDrivers}, {numDrivers, 1}]?

Reply:
Good suggestion. I will consider to try to use a loop to handle all the cases later when integrating with spilling. Thanks!